design/authn_azure/authn_azure_solution_design.md

```diff
@@ -0,0 +1,373 @@
+
+#### Error handling
+
+|    | Given                                                                                     | When                                                                | Then                                                                                                                                         | Manual / UT / Integration | Status     |
+|----|-------------------------------------------------------------------------------------------|---------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|---------------------------|------------|
+| 1  | authn-azure is not whitelisted in ENV or enabled in DB                                    | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (AuthenticatorNotWhitelisted) is returned and logged   | Integration               |            |
+| 2  | Policy is loaded w/o "provider-uri" variable or value is missing                          | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (RoleNotAuthorizedOnWebservice) is returned and logged | Integration               |            |
+| 3  | Policy is loaded w/o defining authenticator webservice                                    | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (WebserviceNotFound) is returned and logged            | Integration               |            |
+| 4  | An incorrectly/undefined provider-uri (Conjur couldn't connect to provider)               | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (ProviderDiscoveryTimeout) is returned and logged      | Integration               |            |
+| 5  | Conjur can make contact with provider but couldn't confirm token signature                | VM sends authentication request to Conjur with invalid Azure token  | 500 Internal Server Error response with an error body with the relevant error message (ProviderTokenInvalid) is returned and logged          | Integration               |            |
+| 6  | One or more annotations are missing from the host policy **                               | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (MissingRequestParam) is returned and logged           | Integration               |            |
```

Comment:
I disagree with the `MissingRequestParam` error. The request itself is fine but the missing param is in the host policy definition. We should return an 401 Unauthorized response in this case and log which annotation is missing.

Reply:
I am confused by your note here. Are we checking the conjur policy being sent in request body or the actual request being sent by the Azure resource. Ex: the resource sends Azure token without "token"

Reply:
If i understand correctly the scenario you described, then the host was loaded without all the required annotations. Then an authentication request is sent with a valid Azure token. So the request itself is valid so `MissingRequestParam` seems wrong.

Reply:
So here I think this should be 2 separate tests,
1. Check if the 3 annotations we expect to be in Host policy are in fact there (RoleMissingAnnotations)
2. Ensure Azure access token has the fields we expect MissingRequestParam

Reply:
1. I agree - this test should be in `ValidateApplicationIdentity`. I would name the error with something related to Application Identity just so it's clear what the annotations are.
2. Not quite. `MissingRequestParam` should be raised if the request body is lacking the `token` field, or if it is empty. If the token exists but doesn't have the `xms_mirid` field it should raise something similar to this:
```
TokenFieldNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
          msg: "Field '{0-field-name}' not found or empty in token",
          code: "CONJ00013E"
        )
```